DOC small note and crossref on the equivalency of recall (macro average) and balanced_accuracy_score

flipz357 opened the pull request:
#### What does this implement/fix? Explain your changes.

It adds a note and cross-reference in the documentation about the equivalency of Balanced Accuracy and Recall (when the latter is used with macro average).

#### Any other comments?

It's conceptually the same. I additionally verified it within sklearn -- This runs without fault:

```python
import numpy as np
from sklearn.metrics import balanced_accuracy_score, recall_score

def generate_yx(num_labels):
    y = np.random.randint(0, num_labels, 500)
    x = np.random.randint(0, num_labels, 500)
    return y, x

for _ in range(1000):
    for n in [2, 3, 4, 5]:
        y, x = generate_yx(n)
        r = recall_score(y, x, average="macro")
        b = balanced_accuracy_score(y, x)
        assert r == b
```

Just think it's something useful to highlight in the Docs.

flipz357 committed:
added note about the relationship of balanced_accuracy and recall_score

flipz357 committed:
remove comma

flipz357 committed:
small text format correction

flipz357 committed:
Update _classification.py

flipz357 committed:
minor formatting

flipz357 committed:
fix crosslink, typo

flipz357 committed:
solve linting issues

flipz357 commented:
added a verification of the information in the main comment.

virchan commented on sklearn/metrics/_classification.py:
```diff
@@ -2360,6 +2360,18 @@ def recall_score(
+    ``recall_score(y_true, y_pred, average=macro)`` is equivalent to
+    ``balanced_accuracy_score(y_true, y_pred)``.
+
+    References
+    ----------
+    .. [1] J. Opitz (2024). `"A Closer Look at Classification Evaluation Metrics
+           and a Critical Reflection of Common Evaluation Practice". Transactions
+           of the Association for Computational Linguistics 12 (2024): 820-836
+           <https://doi.org/10.1162/tacl_a_00675>`_.
```
So, at the end of the day, it comes down to tracing the definitions:

$$\mbox{Macro-averaged Recall} := \displaystyle \sum_{i=1}^{n_{\mathrm{classes}}} \mbox{Recall for class $i$} =: \mbox{Balanced Accuracy}$$

However, it's not clear to me how the newly added reference justify this, or at least, provide a rigourious proof on the equivalence of two metrics.

Maybe pinging @glemaitre for a second opinion here.

flipz357 replied:
Thanks @virchan for checking into this. I think the proof that you were looking for is in Eq. 8 in the referenced paper.

Sorry, I guess I maybe should've highlighted this in the opening post.

virchan reviewed (commented):
Otherwise, LGTM!

flipz357 committed:
point to eq. for equivalency, shorten reference

virchan reviewed (commented):
Thank you for clarifying the equivalence between macro-averaged recall and balanced accuracy, and for providing a reference, @flipz357!

I noticed the referenced article is authored by you. While it’s a valid source, could we consider additional established references to ensure a neutral and widely recognised perspective in the documentation?

Looking forward to your thoughts!

flipz357 commented:
Thanks @virchan. It was a good idea of you to look into this more. After searching, I see now that this is also already mentioned in the [User Guide](https://scikit-learn.org/dev/modules/model_evaluation.html#balanced-accuracy-score), but without pointing to the exact derivation of the equivalency.

How about I change the PR as follows:

1. Remove the suggested reference from the DOCS and integrate it into the linked part of the User guide, where there are other supporting references. I also think that the pointer to the Equation is more informative in this context.

2. Keep in the DOCS of balanced accuracy and macro Recall: The note about their equivalency, and simply add a link to the relevant section in the user guide for further reading. I think that keeping a note directly in the DOCS is still useful to avoid confusion that can come from two metrics that are equivalent but have the same name.

I think that this solution can be the most informative for users.

flipz357 committed:
removed references, added pointer and detail to user guide

flipz357 committed:
found a typo, shorten

flipz357 commented:
Took a stab at it and implemented the solution that I proposed above. The reference is now just one of others and is used to support what was actually already said in the User Guide. The DOCs on the other hand now just note about the equivalency and link to the User Guide.

Also spotted and fixed a typo that was already in the User Guide (A fullstop too much).

Happy to hear what you think @virchan !

virchan reviewed (commented):
Thanks for the follow-up, @flipz357! I appreciate your suggestion and think we can refine it further.

The user guide already defines balanced accuracy as the macro-averaged recall:

<img width="472" alt="Screenshot 2024-11-27 184304" src="https://github.com/user-attachments/assets/f3e0e79b-14cc-46f4-9210-d3d0c721c76d">


Given this, I'm not sure adding the [proposed reference](https://doi.org/10.1162/tacl_a_00675) would provide additional value for users-- just my two cents!

However, we can enhance users' understanding by showing the logic explicitly. Since it's essentially a matter of tracing definitions, the argument itself is quite straightforward.

If you could provide the mathematical argument directly in the user guide using the same / similar notations:

<img width="467" alt="Screenshot 2024-11-27 184905" src="https://github.com/user-attachments/assets/53099755-b9f4-441b-8a07-f2235114a22f">


I believe users would find this more helpful than needing to consult a separate document with a different set of notations.

What do you think about this approach?

flipz357 commented:
Yes, I was exactly meaning the part that you highlighted in red :-) 

I think your approach is good, I just need to get a bit more clear on scikit-learn's notation. As of now I am a bit confused about the "_Sample weight_". It means a weight for an instance, right (not a class)? I am currently not sure if the formula that is currently displayed can then be obviously rewritten as "macro Recall". I think the current definition of sample weight (formula above your blue circle) would allow different instances from the same class to have different weight. Every class will have the same weight (just as in macro Recall), but the weights of the individual instances within a class can differ (this is not the case for the most commonly used variant of macro Recall). Or am I overlooking something obvious here?

flipz357 committed:
add macro recall equivalency to user guide and hedge intro statement in notes

flipz357 commented:
@virchan Took a stab at it! Let me hear what you think.

For now, I still left the reference in: I have found that the statement in the introduction of balanced accuracy should be hedged. It is not strictly true that (as the text says) balanced accuracy doesn't lead to inflated performance values on imbalanced datasets. Maybe this is often the case, but, indeed, it's not strictly true. This is also what I wanted to highlight with my work -- it's difficult to make absolute statements on measures, and complementary measures need be assessed. I added a note with an example and guidance. I think it's a useful recommendation in this context and hedges a bit the strong statement in the beginning.

PS: I could successfully build everything, but CircleAI doesn't like something about the formatting. It also doesn't tell me specifics about the errors.

virchan commented on doc/modules/model_evaluation.rst:
```diff
@@ -575,13 +585,18 @@ or *informedness*.
     * Balanced Accuracy as described in [Urbanowicz2015]_: the average of sensitivity and specificity
       is computed for each class and then averaged over total number of classes.
 
+    There are some cases, where balanced accuracy does not avoid inflated performance
+    scores in imbalanced datasets (for example, when minority classes are detected well, but
+    majority classes aren't). In these cases, accuracy can be low, but balanced accuracy can
+    be high. Therefore [Opitz2024]_ recommend inspecting balanced accuracy together with accuracy.
```
What do you think of the following reference?

[https://imbalanced-learn.org/stable/auto_examples/applications/plot_impact_imbalanced_classes.html](https://imbalanced-learn.org/stable/auto_examples/applications/plot_impact_imbalanced_classes.html)

virchan reviewed (commented):
> I just need to get a bit more clear on scikit-learn's notation.

I think you nailed it. I modified your version to generate the following HTML:

<img width="865" alt="Screenshot 2024-11-28 162006" src="https://github.com/user-attachments/assets/04dcbb25-47d7-4cd8-90a6-75429937b880">

I'd appreciate your thoughts on this.

flipz357 commented:
I like your proposed modification @virchan . It makes the equivalence really clear, and also avoids working with the sample weights, that may be a bit complex to understand for users.

flipz357 commented:
Wondering if I should integrate your proposed change in my PR @virchan ? 

That is, I can integrate what is in your screenshot where you wrote the simpler formula and explained it, and bind in your suggested reference to the concrete demonstration example with imbalanced classes. 

Thanks for the suggestions!

virchan commented:
> Wondering if I should integrate your proposed change in my PR @virchan ?
> 
> That is, I can integrate what is in your screenshot where you wrote the simpler formula and explained it, and bind in your suggested reference to the concrete demonstration example with imbalanced classes.
> 
> Thanks for the suggestions!

Please feel free to go ahead with integrating the changes @flipz357! Once the documentation is correctly rendered, we can work together to see if there are additional opportunities to add value for users. Fingers crossed for smooth CI tests!

I’ve also been thinking about the references as well. 

I agree that the scenario you mentioned-- "for example, when minority classes are detected well, but majority classes aren't"-- is an important edge case. This situation can easily be recreated by modifying the [<code>DummyClassifier</code>](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html) in the [imbalanced-learn example](https://imbalanced-learn.org/stable/auto_examples/applications/plot_impact_imbalanced_classes.html) to use <code>strategy="constant"</code>.

That said, I believe the [imbalanced-learn example](https://imbalanced-learn.org/stable/auto_examples/applications/plot_impact_imbalanced_classes.html) already offers a clear and detailed demonstration of how balanced accuracy addresses inflated performance evaluation. Since it’s a recognised resource, it might suffice as the primary reference for clarity and neutrality-- just my two cents!

If the note about exceptions adds value, perhaps it would make sense to incorporate it directly into the [imbalanced-learn example](https://imbalanced-learn.org/stable/auto_examples/applications/plot_impact_imbalanced_classes.html). Given that your suggestion aligns closely with the example's core thesis, you could consider submitting a new PR to update and improve it.

Also, the imbalanced-learn team overlaps with scikit-learn, so referencing this PR for context could help ensure continuity in the review process.

Looking forward to hearing your thoughts!

flipz357 commented:
Thanks for your thoughts @virchan! I guess then it's best if I close this PR and make a new, clean one from scratch, without the reference, with your suggestion of simplification of formula, and also open one in the imbalanced repo for code demonstration, hopefully all with smooth CI tests. This is a little bit of work, but I'll look into it when I have some time.
